docs: state management

dummdidumm opened the pull request:
- docs about state management
- more helpful error message when trying to run a SvelteKit store on the server

closes #8524
closes #8302
related to #8614

The docs are WIP, I'd like to add more stuff to it.
- maybe move the explanation about cookies from `Loading Data` in there?
- touch on invalidation as a method to handle state
- in general a bit more indepth about the global gotchas, maybe more code examples?
- other thoughts?

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

stephane-vanraes commented on documentation/docs/20-core-concepts/60-state-management.md:
```diff
@@ -0,0 +1,53 @@
+
+- if you need to access the user state only inside `load` functions, use `locals` in server `load` functions
+- if you need to persist the user state across reloads, but only need to access it inside `load` functions, use `cookies` in server `load` functions
+- if you need to access and update the state inside components, use Svelte's [context feature](https://svelte.dev/docs#run-time-svelte-setcontext). That way, the state is scoped to components, which means they are not shared across different requests on the server. The drawback is that you can only set and subscribe to the store at component initialization. SvelteKit's stores from `$app/stores` f ...
```
Could it be useful to have an example of this here ?
I am not sure myself what is meant with this, but are you referring to returning data from `load`, placing it in a store and then use `setContext` in the page to make it available ?

dummdidumm replied:
yes this would be the way - and I agree this probably needs a code example

benmccann commented on documentation/docs/20-core-concepts/60-state-management.md:
```diff
@@ -0,0 +1,53 @@
+How state can be represented:
+- JavaScript variable: Not visible from the outside
+- URL: What page you're on, but also information for the page itself through query parameters for example
+- Cookies: Mostly modifiable through the server, contains for example information about the current user. Can be used to persist state across browser tabs or reloads.
```
```suggestion
- [Cookies](/docs/load#cookies-and-headers): Mostly modifiable through the server. Can be used to persist state across browser tabs or reloads. May contain, for example, information about the current user.
```

benmccann commented on documentation/docs/20-core-concepts/60-state-management.md:
```diff
@@ -0,0 +1,53 @@
+If you are using SSR, the `load` function will run on the server initially, which means that the whole server instance which serves _all_ requests from _all_ users has its `user` state set to the one just requested from the API. To scope this to a single user, you have a couple of options:
+
+- if you need to access the user state only inside `load` functions, use `locals` in server `load` functions
+- if you need to persist the user state across reloads, but only need to access it inside `load` functions, use `cookies` in server `load` functions
```
```suggestion
- if you need to persist the user state across reloads, but only need to access it inside `load` functions, use [`cookies` in server `load` functions](/docs/load#cookies-and-headers)
```

benmccann reviewed (commented):
> maybe move the explanation about cookies from Loading Data in there?

I think the cookies docs make sense in Loading Data where they are now, but we should certainly link to them

Rich-Harris commented:
Agree that we should clarify why you can't subscribe to app stores outside component initialization (though the improved error messages are the real win here), but I do think we should be careful about how far into the weeds we go. For example the 'types of state' section feels a bit intimidating ("I need to understand all these different kinds of things? I just wanted to load some data") and yet is arguably incomplete (it doesn't mention stores, or how API/backend relates to page data). 

There's a cost to adding documentation — it makes it harder for a newcomer to RTFM — and so we need to be sure that we're solving a real problem. #8524 is painfully vague about our docs' shortcomings, to the point that I'm not sure _any_ PR could claim to close it.

benmccann commented:
I agree the docs in this PR are too broad. It seems like the thing that people trip over a lot is how to use stores with SSR. I think we could cover that much more concisely. Perhaps we could document the pattern Rich recommended in https://github.com/sveltejs/kit/issues/7105#issuecomment-1305856190 in the load docs or something

dummdidumm commented:
I'm good with removing the broader section, it was intended as a start to a more full-fledged docs page - imagine about twice the content there is now in the end. But I don't have a good vision of how it all could look like.

We already have a short section on stores and SSR in the load docs, which noone reads because noone thinks that it's hidden in there. I think a general "State Management" page makes sense since it's a widely used term so people are far more likely to click on that to see how things are done "the SvelteKit way".

stephane-vanraes commented:
Perhaps the docs should simple focus on the "API" instead.
And in addition you make a `guides` segment that goes a bit deeper in certain parts like state management.
NextJS has for example a more _in detail_ section on how to make forms https://nextjs.org/docs/guides/building-forms
Remix also has a "Guides" section in their docs that cover things like data loading in more detail.

dummdidumm commented:
I update the doc to remove that "types of state" prelude and added a section on form interaction instead. It has reminiscents of a trouble shooting or "do it this way instead" which I think is a good addition.

benmccann commented on documentation/docs/20-core-concepts/60-state-management.md:
```diff
@@ -0,0 +1,94 @@
+
+## Careful with global state
+
+If you are creating an [SPA](/docs/glossary#csr) with SvelteKit, you can create global state freely, as you can be sure that it's only initialized inside the user's browser. If you use [SSR](/docs/glossary#ssr) however, you have to watch out for a couple of things when managing state. In many server environments, a single instance of your app will serve multiple users (this is not specific to Sve ...
```
```suggestion
If you are creating an [SPA](glossary#csr) with SvelteKit, you can create global state freely, as you can be sure that it's only initialized inside the user's browser. If you use [SSR](glossary#ssr) however, you have to watch out for a couple of things when managing state. In many server environments, a single instance of your app will serve multiple users (this is not specific to SvelteKit - it's one of the gotchas of working with such environments). For that reason, per-request or per-user state must not be stored in global variables.
```

tomecko replied:
```suggestion
If you are creating a [SPA](/docs/glossary#csr) with SvelteKit, you can create a global state freely, as you can be sure that it's only initialized inside the user's browser. However, if you use [SSR](/docs/glossary#ssr), you have to watch out for a couple of things when managing state. In many server environments, a single instance of your app will serve multiple users (this is not specific to SvelteKit - it's one of the gotchas of working with such environments). For that reason, per-request or per-user state must not be stored in global variables.
```

benmccann commented on documentation/docs/20-core-concepts/60-state-management.md:
```diff
@@ -0,0 +1,94 @@
+
+Managing state is one of the hardest parts in application development. This section covers various use cases with regards to state management and what to watch out for.
+
+## Careful with global state
```
```suggestion
## Avoid global state in SSR
```

dummdidumm replied:
Is "avoid" a too strong word, given that it's ok to have global state if you make sure to not update it to request-specific values during ssr?

benmccann replied:
I tend think it'd be okay as we can go into greater detail in the text and avoid just means "try not to use" and not "do not use at all"

benmccann commented on documentation/docs/20-core-concepts/60-state-management.md:
```diff
@@ -0,0 +1,94 @@
+
+## Managing forms
+
+If you are coming from regular Svelte, you may be used to do something like this when creating a form to react to user input:
```
```suggestion
If you are coming from regular Svelte, you may be used to doing something like this when creating a form to react to user input:
```

benmccann reviewed (commented):
The forms stuff feels out-of-place in the state section to me. It seems like it belongs much better on the form actions page.

We've really got to get the PR doc deploys fixed :smile:

dummdidumm commented:
I was debating this myself, there's duplication to the form actions section, but if you browse the docs just from looking at the titles, you may end up here searching for the best way to do this. It's a little like [the error docs](https://kit.svelte.dev/docs/errors) which also contain references to various other places in the docs.

tomecko commented on packages/kit/src/runtime/app/stores.js:
```diff
@@ -55,3 +55,19 @@ export const updated = {
+		return getStores()[name];
+	} catch (e) {
+		throw new Error(
+			`Store '${name}' is not available outside of a Svelte component on the server, as its bound to component context there. ` +
```
```suggestion
			`Store '${name}' is not available outside of a Svelte component on the server, as it's bound to component context there. ` +
```

tomecko commented on packages/kit/src/runtime/app/stores.js:
```diff
@@ -55,3 +55,19 @@ export const updated = {
+	} catch (e) {
+		throw new Error(
+			`Store '${name}' is not available outside of a Svelte component on the server, as its bound to component context there. ` +
+				'For more information, see https://kit.svelte.dev/docs/state-management#careful-with-global-state'
```
heads up to update the link if there is change based on this thread: https://github.com/sveltejs/kit/pull/8547/files#r1085676189

tomecko commented on documentation/docs/20-core-concepts/60-state-management.md:
```diff
@@ -0,0 +1,94 @@
+---
```
should the filename rather start with `50`?

dummdidumm commented:
I moved most of the details around form actions to its respective docs page, but kept a reference to it so it's more discoverable. I also added a short URL section.

benmccann commented on documentation/docs/20-core-concepts/50-state-management.md:
```diff
@@ -0,0 +1,46 @@
+
+## Avoid global state in SSR
+
+If you are creating an [SPA](glossary#csr) with SvelteKit, you can create global state freely, as you can be sure that it's only initialized inside the user's browser. If you use [SSR](glossary#ssr) however, you have to watch out for a couple of things when managing state. In many server environments, a single instance of your app will serve multiple users (this is not specific to SvelteKit - it' ...
```
I read SPA as the word "spa", so "an spa" sounds weird to me - though it'd be correct for folks reading it as s-p-a. Maybe we just write out "single page application" to avoid the issue
```suggestion
If you are creating a [single page application (SPA)](glossary#csr) with SvelteKit, you can create global state freely, as you can be sure that it's only initialized inside the user's browser. If you use [SSR](glossary#ssr) however, you have to watch out for a couple of things when managing state. In many server environments, a single instance of your app will serve multiple users (this is not specific to SvelteKit - it's one of the gotchas of working with such environments). For that reason, per-request or per-user state must not be stored in global variables.
```

dummdidumm replied:
I linked there so people can look up good to enter spa mode

benmccann replied:
doesn't `glossary#spa` better describe how to enter SPA mode than the CSR entry does? Neither says directly, but the SPA entry directs users to `adapter-static#spa-mode`

benmccann commented on documentation/docs/20-core-concepts/50-state-management.md:
```diff
@@ -0,0 +1,46 @@
+If you are using SSR, the `load` function will run on the server initially, which means that the whole server instance which serves _all_ requests from _all_ users has its `user` state set to the one just requested from the API. To scope this to a single user, you have a couple of options:
+
+- if you need to access the user state only inside server `load` functions, use [`locals`](hooks#server-hooks-handle)
+- if you need to persist the user state across reloads, but only need to access it inside `load` functions, use [`cookies` in server `load` functions](load#cookies-and-headers)
```
cookies aren't available only in load functions, but also in api endpoints, etc. also, you'd often not store the state in a cookie, but a key to lookup the state

benmccann commented on documentation/docs/20-core-concepts/50-state-management.md:
```diff
@@ -0,0 +1,46 @@
+
+- if you need to access the user state only inside server `load` functions, use [`locals`](hooks#server-hooks-handle)
+- if you need to persist the user state across reloads, but only need to access it inside `load` functions, use [`cookies` in server `load` functions](load#cookies-and-headers)
+- if you need to access and update the state inside components, use Svelte's [context feature](https://svelte.dev/docs#run-time-svelte-setcontext). That way, the state is scoped to components, which means they are not shared across different requests on the server. The drawback is that you can only set and subscribe to the store at component initialization. SvelteKit's stores from `$app/stores` f ...
```
users may not know what "component initialization" means. maybe we can clarify where it can be used (e.g. in a `<script>` tag, but not reactive statement - I don't know if that's exactly right but just to give an idea)

HarryAllen1 commented on documentation/docs/20-core-concepts/20-load.md:
```diff
@@ -162,6 +162,14 @@ As we've seen, there are two types of `load` function:
+
+Server `load` functions _always_ run on the server.
+
+By default universal `load` functions run on the server during SSR when the user first visits your page. They will then run again during hydration, reusing any responses from [fetch requests](#making-fetch-requests). All subsequent invocations of universal `load` functions happen in the browser. You can customize the behavior through [page options](page-options). If you disable [server side rende ...
```
```suggestion
By default, universal `load` functions run on the server during SSR when the user first visits your page. They will then run again during hydration, reusing any responses from [fetch requests](#making-fetch-requests). All subsequent invocations of universal `load` functions happen in the browser. You can customize the behavior through [page options](page-options). If you disable [server side rendering](page-options#ssr), you'll get an SPA and universal `load` functions _always_ run on the client.
```

Missing comma here

Rich-Harris commented on documentation/docs/20-core-concepts/20-load.md:
```diff
@@ -162,6 +162,14 @@ As we've seen, there are two types of `load` function:
 
 Conceptually, they're the same thing, but there are some important differences to be aware of.
 
+### When does which load function run
```
```suggestion
### When does which load function run?
```

Rich-Harris commented on documentation/docs/20-core-concepts/50-state-management.md:
```diff
@@ -0,0 +1,75 @@
+---
+title: State Management
```
```suggestion
title: State management
```

Rich-Harris commented on documentation/docs/20-core-concepts/50-state-management.md:
```diff
@@ -0,0 +1,75 @@
+
+If you have global data whose initial state is not dependent on a request (in other words, it's always the same), then you can keep storing that data globally, as long as you make sure you don't update it during the initial rendering on the server (during load or component render).
+
+## Managing forms
+
+When coming from a pure Svelte or JavaScript background, you might be used to handling all form interactions through JavaScript. This works well when JavaScript is available but results in unresponsive UI when it isn't (which may be [more often than you think](https://kryogenix.org/code/browser/everyonehasjs.html)). If this is a concern to you, leverage SvelteKit's [form actions](form-actions) in ...
```
Does this belong here? It's well covered by the form section, and I don't think of mutations as part of state management

```suggestion
```

dummdidumm replied:
Mutating state is an integral part of managing state to me. If all state was just readonly, managing it would be a lot easier 😄

Rich-Harris commented:
#9239
